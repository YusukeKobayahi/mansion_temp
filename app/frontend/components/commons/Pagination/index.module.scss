@import "~/styles/variables";
@import "~/styles/utils";

.container {
  margin-left: auto;

  @include screenMd {
    width: 100%;
  }
}

.pagination {
  @include screenMd {
    margin-top: 12px;
    width: 100%;
  }

  &Inner {
    display: flex;

    span {
      position: relative;
      border: 1px solid #e5e5e5;
      width: 50px;
      line-height: 50px;
      text-align: center;
      font-size: 24px;

      @include screenMd {
        max-width: 50px;
        width: unset;
        flex: 1;
      }

      &.current {
        color: #fcf9f2;
        background-color: #bca565;
      }

      &.disable {
        background-color: #f4f4f4;
      }

      button {
        display: block;
        color: #bca565;
        width: 100%;
        height: 100%;
        padding: 0;
        background-color: unset;
      }

      &:nth-child(1) {
        border-radius: 4px 0 0 4px;

        &::after {
          content: "";
          width: 10px;
          height: 10px;
          border-top: 2px solid #bca565;
          border-left: 2px solid #bca565;
          position: absolute;
          top: 50%;
          left: 45%;
          transform: translateY(-50%) rotate(-45deg);
        }
      }

      &:nth-last-child(1) {
        border-radius: 0 4px 4px 0;

        &::after {
          content: "";
          width: 10px;
          height: 10px;
          border-top: 2px solid #bca565;
          border-right: 2px solid #bca565;
          position: absolute;
          top: 50%;
          left: 35%;
          transform: translateY(-50%) rotate(45deg);
        }
      }
    }
  }
}
