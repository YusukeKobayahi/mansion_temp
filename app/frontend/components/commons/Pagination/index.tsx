import Link from "next/link";
import styles from "~/components/commons/Pagination/index.module.scss";
import { QueryVariables } from "~/lib/types";
import { ModelQueryKeys } from "~/lib/property";
import { buildHrefByIsNoindex } from "~/lib/utils";
import omitBy from "lodash/omitBy";
import includes from "lodash/includes";
import { event } from "~/lib/gtag";

export interface PaginationProps {
  currentPage: number;
  totalPages: number;
  pathname: string;
  searchCondition: QueryVariables;
}

const Pagination: React.FC<PaginationProps> = ({
  currentPage = 1,
  totalPages,
  pathname,
  searchCondition,
}) => {
  const query = omitBy(
    { ...searchCondition },
    (value, key) =>
      includes(ModelQueryKeys, key) && String(value).split(",").length === 1
  );
  const list = Array.from({ length: totalPages }, (_, i) => i + 1);
  const index = currentPage - 1;
  const pagination: number[] = [];
  const setIndex = (n: number) => {
    return n < 0 || n > totalPages - 1 ? -1 : n;
  };
  const length = 5;
  if (list && list.length > 0) {
    if (list.length < 5) pagination.push(...list);
    else {
      pagination.push(list[index]);
      for (let i = 1; pagination.length < length; i++) {
        if (list[setIndex(index - i)])
          pagination.unshift(list[setIndex(index - i)]);
        if (list[setIndex(index + i)])
          pagination.push(list[setIndex(index + i)]);
      }
    }
  }
  const clickPagination = () => {
    event({
      action: "msl_page_click",
      category: "msl_ichiran",
      label: "msl_pagination",
      value: 1,
    });
  };
  return (
    <section className={styles.container}>
      {totalPages !== 0 && (
        <div className={styles.pagination}>
          <div className={styles.paginationInner}>
            {currentPage !== 1 ? (
              <span className={styles.able}>
                <Link
                  href={buildHrefByIsNoindex(
                    pathname,
                    Object.assign({}, query, { page: currentPage - 1 })
                  )}
                  prefetch={false}
                  passHref
                >
                  <button onClick={clickPagination}>
                    <b></b>
                  </button>
                </Link>
              </span>
            ) : (
              <span className={styles.disable}></span>
            )}
            {pagination.map((page: number, index) => {
              const inner =
                page === currentPage ? (
                  <span className={styles.current} key={index}>
                    {page}
                  </span>
                ) : (
                  <span className={styles.able} key={index}>
                    <Link
                      href={buildHrefByIsNoindex(
                        pathname,
                        Object.assign({}, query, { page })
                      )}
                      prefetch={false}
                      passHref
                    >
                      <button className={styles.able} onClick={clickPagination}>
                        {page}
                      </button>
                    </Link>
                  </span>
                );
              return inner;
            })}
            {currentPage !== totalPages ? (
              <span className={styles.able}>
                <Link
                  href={buildHrefByIsNoindex(
                    pathname,
                    Object.assign({}, query, { page: currentPage + 1 })
                  )}
                  passHref
                >
                  <button onClick={clickPagination}>
                    <b></b>
                  </button>
                </Link>
              </span>
            ) : (
              <span className={styles.disable}></span>
            )}
          </div>
        </div>
      )}
    </section>
  );
};

export default Pagination;
