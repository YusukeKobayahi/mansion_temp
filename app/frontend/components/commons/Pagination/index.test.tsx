import { render } from "@testing-library/react";
import Pagination from "~/components/commons/Pagination";

describe("components/commons/Pagination", () => {
  it("正常に表示される", () => {
    const dummyData = {
      currentPage: 5,
      totalPages: 10,
      pathname: "/cities/13101/",
    };
    const searchCondition = {};
    const { asFragment } = render(
      <Pagination {...dummyData} searchCondition={searchCondition} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
