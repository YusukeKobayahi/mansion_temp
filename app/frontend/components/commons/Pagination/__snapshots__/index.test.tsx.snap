// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/commons/Pagination 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="pagination"
    >
      <div
        class="paginationInner"
      >
        <span
          class="able"
        >
          <button
            href="/cities/13101?page=4"
          >
            <b />
          </button>
        </span>
        <span
          class="able"
        >
          <button
            class="able"
            href="/cities/13101?page=3"
          >
            3
          </button>
        </span>
        <span
          class="able"
        >
          <button
            class="able"
            href="/cities/13101?page=4"
          >
            4
          </button>
        </span>
        <span
          class="current"
        >
          5
        </span>
        <span
          class="able"
        >
          <button
            class="able"
            href="/cities/13101?page=6"
          >
            6
          </button>
        </span>
        <span
          class="able"
        >
          <button
            class="able"
            href="/cities/13101?page=7"
          >
            7
          </button>
        </span>
        <span
          class="able"
        >
          <button
            href="/cities/13101?page=6"
          >
            <b />
          </button>
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
